apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).SetResult(true), tcs);
+            return tcs.Task;
+        }
+
+        private static async Task StartSessionCreating(CancellationToken cancellationToken)
+        {
+            lock (s_waitSync)
+            {
+                if (CurrentActiveSessions < MaximumActiveSessions)
+                {
+                    s_sessionsCreating++;
+                    return;
+                }
+            }
+            var canceledTask = cancellationToken.WhenCanceled();
+            //we need to block or throw because we are out of allocations.
+            while (true)
+            {
+                if (WaitOnResourcesExhausted)
+                {
+                    TaskCompletionSource<int> signal = new TaskCompletionSource<int>();
+                    s_waitQueue.Enqueue(signal);
+                    await Task.WhenAny(signal.Task, canceledTask).ConfigureAwait(false);
+                    cancellationToken.ThrowIfCancellationRequested();
+                    lock (s_waitSync)
+                    {
+                        if (CurrentActiveSessions < MaximumActiveSessions)
+                        {
+                            s_sessionsCreating++;
+                            return;
+                        }
+                    }
+                }
+                else
+                {
+                    throw new RpcException(new Status(StatusCode.ResourceExhausted, "Number of available Sessions exhausted."));
+                }
+            }
+        }
+
+        private static void EndSessionCreating()
+        {
+            lock (s_waitSync)
+            {
+                s_sessionsCreating--;
+            }
+            SignalAnyWaitingRequests();
+        }
+
+        /// <summary>
+        /// Allocates a session from the pool if possible, or creates a new Spanner Session.
+        /// </summary>
+        /// <param name="spannerClient"></param>
+        /// <param name="project"></param>
+        /// <param name="spannerInstance"></param>
+        /// <param name="spannerDatabase"></param>
+        /// <param name="options"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public static async Task<Session> CreateSessionFromPoolAsync(this SpannerClient spannerClient,
+            string project, string spannerInstance, string spannerDatabase, TransactionOptions options, CancellationToken cancellationToken)
+        {
+            project.ThrowIfNullOrEmpty(nameof(project));
+            spannerInstance.ThrowIfNullOrEmpty(nameof(spannerInstance));
+            spannerDatabase.ThrowIfNullOrEmpty(nameof(spannerDatabase));
+            
+            await StartSessionCreating(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                var sessionPoolKey = new SessionPoolKey {
```

Comment:
When SessionPoolKey is immutable, these will just be constructor calls.
(I thought I'd seen this change before, so I'm confused by this...)